src/Compilers/Core/Portable/Text/CompositeText.cs

```diff
@@ -373,5 +378,131 @@ private static void TrimInaccessibleText(ArrayBuilder<SourceText> segments)
                 segments.Add(writer.ToSourceText());
             }
         }
+
+        /// <summary>
+        /// Delegates to SourceTexts within the CompositeText to determine line information.
+        /// </summary>
+        private sealed class CompositeLineInfo : TextLineCollection
+        {
+            private readonly CompositeText _compositeText;
+
+            // The starting line number for the correspondingly indexed SourceTexts in _compositeText.Segments
+            private readonly int[] _segmentLineNumbers;
+
+            // The total number of lines in our _compositeText
+            private readonly int _lineCount;
+
+            public CompositeLineInfo(CompositeText compositeText)
+            {
+                _compositeText = compositeText;
+                _segmentLineNumbers = new int[compositeText.Segments.Length];
+
+                var accumulatedLineCount = 0;
+                for (int i = 0; i < compositeText.Segments.Length; i++)
+                {
+                    _segmentLineNumbers[i] = accumulatedLineCount;
```

Comment:
ok.  might be good with comments.  it's good to understand that this is initially zero because, of course, the first segment must start at zero.